openlcs/packages/mixins.py

```diff
@@ -254,20 +200,64 @@ def save_scan_result(self, **kwargs):
             licenses = kwargs.pop('licenses')
             data = licenses.get('data')
             if not licenses.get('has_exception'):
-                with transaction.atomic():
-                    # Exist same files in different paths.
-                    self.save_file_license_scan(list(set(file_ids)))
-                    self.save_license_detections(path_file_dict, data)
-                    self.update_scan_flag(source, "license_scan")
+                license_detector = settings.LICENSE_SCANNER
+                filters = [Q(file__in=file_ids), Q(detector=license_detector)]
+
+                # Retry 5 times to bypass possible concurrency issue.
+                max_retries = settings.SAVE_DATA_MAX_RETRIES
+                for i in range(max_retries):
+                    try:
+                        license_file_ids = FileLicenseScan.objects.filter(
+                            *filters).values_list('file_id', flat=True)
+                        new_file_ids = list(
+                            set(file_ids).difference(license_file_ids))
+                        with transaction.atomic():
+                            # Exist same files in different paths.
+                            self.save_file_license_scan(
+                                new_file_ids, license_detector)
+                            self.save_license_detections(
+                                path_file_dict, data, license_detector)
+                            self.update_scan_flag(source, "license_scan")
+                    except IntegrityError as err:
+                        if i == max_retries - 1:
```

Comment:
nitpick: suppose `SAVE_DATA_MAX_RETRIES=2`, this "==" means the forloop runs only once.

Reply:
`i` will 0, 1 I think.

Reply:
you want to retry twice in case MAX_RETRIES=2

Reply:
Yes, retry twice, the same as setting `MAX_RETRIES`